File: dotnet/cs/src/STM.cs
/*
 * File: STM.cs
 * Project: src
 * Created Date: 20/08/2023
 * Author: Shun Suzuki
 * -----
 * Last Modified: 06/12/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2023 Shun Suzuki. All rights reserved.
 * 
 */

#if UNITY_2018_3_OR_NEWER
#define USE_SINGLE
#endif

using System;
using System.Collections.Generic;
using System.Linq;

using AUTD3Sharp.Internal;
using AUTD3Sharp.NativeMethods;

#if UNITY_2020_2_OR_NEWER
#nullable enable
#endif

#if UNITY_2018_3_OR_NEWER
using UnityEngine;
using Vector3 = UnityEngine.Vector3;
#else
using Vector3 = AUTD3Sharp.Utils.Vector3d;
#endif

#if USE_SINGLE
using float_t = System.Single;
#else
using float_t = System.Double;
#endif

namespace AUTD3Sharp
{
    namespace STM
    {
        public abstract class STM : IDatagram
        {
            private readonly float_t? _freq;
            private readonly TimeSpan? _period;
            private readonly SamplingConfiguration? _samplingConfig;
            protected int StartIdxV;
            protected int FinishIdxV;

            protected STM(float_t? freq, TimeSpan? period, SamplingConfiguration? config)
            {
                _freq = freq;
                _period = period;
                _samplingConfig = config;
                StartIdxV = -1;
                FinishIdxV = -1;
            }

            DatagramPtr IDatagram.Ptr(Geometry geometry) => STMPtr(geometry);

            internal abstract DatagramPtr STMPtr(Geometry geometry);

            public ushort? StartIdx => StartIdxV == -1 ? null : (ushort?)StartIdxV;

            public ushort? FinishIdx => FinishIdxV == -1 ? null : (ushort?)FinishIdxV;

            internal STMPropsPtr Props()
            {
                var ptr = new STMPropsPtr();
                if (_freq != null)
                    ptr = NativeMethodsBase.AUTDSTMPropsNew(_freq.Value);
                if (_period != null)
                    ptr = NativeMethodsBase.AUTDSTMPropsFromPeriod((ulong)(_period.Value.TotalSeconds * 1000 * 1000 *
                                                                              1000));
                if (_samplingConfig != null)
                    ptr = NativeMethodsBase.AUTDSTMPropsFromSamplingConfig(_samplingConfig.Value.Internal);
                ptr = NativeMethodsBase.AUTDSTMPropsWithStartIdx(ptr, StartIdxV);
                ptr = NativeMethodsBase.AUTDSTMPropsWithFinishIdx(ptr, FinishIdxV);
                return ptr;
            }

            protected float_t FreqFromSize(int size) => NativeMethodsBase.AUTDSTMPropsFrequency(Props(), (ulong)size);

            protected TimeSpan PeriodFromSize(int size) =>
                TimeSpan.FromSeconds(NativeMethodsBase.AUTDSTMPropsPeriod(Props(), (ulong)size) / 1000.0 / 1000.0 /
                                     1000.0);

            protected SamplingConfiguration SamplingConfigFromSize(int size) => new SamplingConfiguration(
                NativeMethodsBase.AUTDSTMPropsSamplingConfig(Props(), (ulong)size).Validate());
        }

        /// <summary>
        /// FocusSTM is an STM for moving a focal point
        /// </summary>
        /// <remarks>
        /// <para>The sampling timing is determined by hardware, thus the sampling time is precise.</para>
        /// <para>FocusSTM has following restrictions:</para>
        /// <list>
        /// <item>The maximum number of sampling points is 65536.</item>
        /// <item>The sampling frequency is <see cref="AUTD3.FPGAClkFreq">AUTD3.FPGAClkFreq</see>/N, where `N` is a 32-bit unsigned integer and must be at 4096.</item>
        /// </list></remarks>
        public sealed class FocusSTM : STM
        {
            private readonly List<float_t> _points;
            private readonly List<EmitIntensity> _intensities;

            private FocusSTM(float_t? freq, TimeSpan? period, SamplingConfiguration? config) : base(
                freq, period, config)
            {
                _points = new List<float_t>();
                _intensities = new List<EmitIntensity>();
            }

            public FocusSTM(float_t freq) : this(freq, null, null)
            {
            }

            public static FocusSTM FromPeriod(TimeSpan period)
            {
                return new FocusSTM(null, period, null);
            }

            public static FocusSTM FromSamplingConfig(SamplingConfiguration config)
            {
                return new FocusSTM(null, null, config);
            }

            /// <summary>
            /// Add focus point
            /// </summary>
            /// <param name="point">Focus point</param>
            /// <param name="intensity">Emission intensity</param>
            /// <returns></returns>
            public FocusSTM AddFocus(Vector3 point, EmitIntensity? intensity = null)
            {
                _points.Add(point.x);
                _points.Add(point.y);
                _points.Add(point.z);
                _intensities.Add(intensity ?? EmitIntensity.Max);
                return this;
            }

            /// <summary>
            /// Add foci
            /// </summary>
            /// <param name="iter">Enumerable of foci</param>
            public FocusSTM AddFociFromIter(IEnumerable<Vector3> iter)
            {
                return iter.Aggregate(this, (stm, point) => stm.AddFocus(point));
            }

            /// <summary>
            /// Add foci
            /// </summary>
            /// <param name="iter">Enumerable of foci and duty shifts</param>
            public FocusSTM AddFociFromIter(IEnumerable<(Vector3, EmitIntensity)> iter)
            {
                return iter.Aggregate(this, (stm, point) => stm.AddFocus(point.Item1, point.Item2));
            }

            public FocusSTM WithStartIdx(ushort? startIdx)
            {
                StartIdxV = startIdx ?? -1;
                return this;
            }

            public FocusSTM WithFinishIdx(ushort? finishIdx)
            {
                FinishIdxV = finishIdx ?? -1;
                return this;
            }

            public float_t Frequency => FreqFromSize(_intensities.Count);
            public TimeSpan Period => PeriodFromSize(_intensities.Count);
            public SamplingConfiguration SamplingConfiguration => SamplingConfigFromSize(_intensities.Count);

            internal override DatagramPtr STMPtr(Geometry geometry)
            {
                var points = _points.ToArray();
                var intensities = _intensities.ToArray();
                unsafe
                {
                    fixed (float_t* pp = &points[0])
                    fixed (EmitIntensity* ps = &intensities[0])
                        return NativeMethodsBase.AUTDSTMFocus(Props(), pp, (byte*)ps, (ulong)_intensities.Count)
                            .Validate();
                }
            }
        }

        /// <summary>
        /// FocusSTM is an STM for moving Gains
        /// </summary>
        /// <remarks>
        /// <para>The sampling timing is determined by hardware, thus the sampling time is precise.</para>
        /// <para>FocusSTM has following restrictions:</para>
        /// <list>
        /// <item>The maximum number of sampling Gain is 2048.</item>
        /// <item>The sampling frequency is <see cref="AUTD3.FPGAClkFreq">AUTD3.FPGAClkFreq</see>/N, where `N` is a 32-bit unsigned integer and must be at 4096.</item>
        /// </list></remarks>
        public sealed class GainSTM : STM
        {
            private readonly List<Internal.Gain> _gains;
            private GainSTMMode _mode;

            private GainSTM(float_t? freq, TimeSpan? period, SamplingConfiguration? config) : base(
                freq, period, config)
            {
                _gains = new List<Internal.Gain>();
                _mode = GainSTMMode.PhaseIntensityFull;
            }

            public GainSTM(float_t freq) : this(freq, null, null)
            {
            }

            public static GainSTM FromPeriod(TimeSpan period)
            {
                return new GainSTM(null, period, null);
            }

            public static GainSTM FromSamplingConfig(SamplingConfiguration config)
            {
                return new GainSTM(null, null, config);
            }

            /// <summary>
            /// Add Gain
            /// </summary>
            /// <param name="gain">Gain</param>
            /// <returns></returns>
            public GainSTM AddGain(Internal.Gain gain)
            {
                _gains.Add(gain);
                return this;
            }

            /// <summary>
            /// Add Gains
            /// </summary>
            /// <param name="iter">Enumerable of Gains</param>
            public GainSTM AddGainsFromIter(IEnumerable<Internal.Gain> iter)
            {
                return iter.Aggregate(this, (stm, gain) => stm.AddGain(gain));
            }

            public GainSTM WithStartIdx(ushort? startIdx)
            {
                StartIdxV = startIdx ?? -1;
                return this;
            }

            public GainSTM WithFinishIdx(ushort? finishIdx)
            {
                FinishIdxV = finishIdx ?? -1;
                return this;
            }

            public GainSTM WithMode(GainSTMMode mode)
            {
                _mode = mode;
                return this;
            }

            public float_t Frequency => FreqFromSize(_gains.Count);
            public TimeSpan Period => PeriodFromSize(_gains.Count);
            public SamplingConfiguration SamplingConfiguration => SamplingConfigFromSize(_gains.Count);


            internal override DatagramPtr STMPtr(Geometry geometry)
            {
                var gains = _gains.Select(g => g.GainPtr(geometry)).ToArray();
                unsafe
                {
                    fixed (GainPtr* gp = &gains[0])
                        return NativeMethodsBase.AUTDSTMGain(Props(), gp, (uint)gains.Length, _mode).Validate();
                }
            }
        }
    }
}

#if UNITY_2020_2_OR_NEWER
#nullable restore
#endif

File: cpp/tests/internal/stm.cpp
// File: stm.cpp
// Project: internal
// Created Date: 26/09/2023
// Author: Shun Suzuki
// -----
// Last Modified: 02/12/2023
// Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
// -----
// Copyright (c) 2023 Shun Suzuki. All rights reserved.
//

#include <gtest/gtest.h>

#include <autd3/gain/focus.hpp>
#include <autd3/gain/uniform.hpp>
#include <autd3/internal/stm.hpp>
#include <ranges>

#include "utils.hpp"

TEST(STMTest, FocusSTM) {
  auto autd = create_controller();

  constexpr double radius = 30.0;
  constexpr int size = 2;
  autd3::internal::Vector3 center = autd.geometry().center() + autd3::internal::Vector3(0, 0, 150);
  auto stm = autd3::internal::FocusSTM(1).add_foci_from_iter(std::views::iota(0) | std::views::take(size) | std::views::transform([&](auto i) {
                                                               const double theta = 2 * autd3::internal::pi * i / size;
                                                               return autd3::internal::ControlPoint{
                                                                   center + autd3::internal::Vector3(radius * cos(theta), radius * sin(theta), 0),
                                                                   autd3::internal::EmitIntensity::maximum()};
                                                             }));
  ASSERT_TRUE(autd.send_async(stm).get());
  for (const auto& dev : autd.geometry()) {
    ASSERT_FALSE(autd.link().is_stm_gain_mode(dev.idx()));
  }

  ASSERT_EQ(1, stm.frequency());
  ASSERT_EQ(2, stm.sampling_config().frequency());
  ASSERT_EQ(10240000u, stm.sampling_config().frequency_division());
  ASSERT_EQ(std::chrono::microseconds(500000), stm.sampling_config().period());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(10240000u, autd.link().stm_frequency_division(dev.idx()));
  }

  ASSERT_EQ(std::nullopt, stm.start_idx());
  ASSERT_EQ(std::nullopt, stm.finish_idx());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(-1, autd.link().stm_start_idx(dev.idx()));
    ASSERT_EQ(-1, autd.link().stm_finish_idx(dev.idx()));
  }

  stm.with_start_idx(0);
  ASSERT_TRUE(autd.send_async(stm).get());
  ASSERT_EQ(0, stm.start_idx());
  ASSERT_EQ(std::nullopt, stm.finish_idx());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(0, autd.link().stm_start_idx(dev.idx()));
    ASSERT_EQ(-1, autd.link().stm_finish_idx(dev.idx()));
  }

  stm.with_start_idx(std::nullopt);
  stm.with_finish_idx(0);
  ASSERT_TRUE(autd.send_async(stm).get());
  ASSERT_EQ(std::nullopt, stm.start_idx());
  ASSERT_EQ(0, stm.finish_idx());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(-1, autd.link().stm_start_idx(dev.idx()));
    ASSERT_EQ(0, autd.link().stm_finish_idx(dev.idx()));
  }

  stm = autd3::internal::FocusSTM::from_sampling_config(autd3::internal::SamplingConfiguration::from_frequency_division(512))
            .add_focus(center)
            .add_focus(center);
  ASSERT_TRUE(autd.send_async(stm).get());
  ASSERT_EQ(20000.0, stm.frequency());
  ASSERT_EQ(2 * 20000.0, stm.sampling_config().frequency());
  ASSERT_EQ(512u, stm.sampling_config().frequency_division());
  ASSERT_EQ(std::chrono::microseconds(25), stm.sampling_config().period());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(512u, autd.link().stm_frequency_division(dev.idx()));
  }

  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(2u, autd.link().stm_cycle(dev.idx()));
    auto [intensities, phases] = autd.link().intensities_and_phases(dev.idx(), 0);
    ASSERT_TRUE(std::ranges::any_of(intensities, [](auto d) { return d != 0; }));
    ASSERT_TRUE(std::ranges::any_of(phases, [](auto p) { return p != 0; }));

    std::tie(intensities, phases) = autd.link().intensities_and_phases(dev.idx(), 1);
    ASSERT_TRUE(std::ranges::any_of(intensities, [](auto d) { return d != 0; }));
    ASSERT_TRUE(std::ranges::any_of(phases, [](auto p) { return p != 0; }));
  }
}

TEST(STMTest, GainSTM) {
  auto autd = autd3::internal::ControllerBuilder()
                  .add_device(autd3::internal::geometry::AUTD3(autd3::internal::Vector3::Zero()))
                  .add_device(autd3::internal::geometry::AUTD3(autd3::internal::Vector3::Zero()))
                  .open_with_async(autd3::link::Audit::builder())
                  .get();

  constexpr double radius = 30.0;
  constexpr int size = 2;
  autd3::internal::Vector3 center = autd.geometry().center() + autd3::internal::Vector3(0, 0, 150);
  auto stm = autd3::internal::GainSTM(1).add_gains_from_iter(std::views::iota(0) | std::views::take(size) | std::views::transform([&](auto i) {
                                                               const double theta = 2 * autd3::internal::pi * i / size;
                                                               return autd3::gain::Focus(
                                                                   center + autd3::internal::Vector3(radius * cos(theta), radius * sin(theta), 0));
                                                             }));
  ASSERT_TRUE(autd.send_async(stm).get());
  for (const auto& dev : autd.geometry()) {
    ASSERT_TRUE(autd.link().is_stm_gain_mode(dev.idx()));
  }

  ASSERT_EQ(1, stm.frequency());
  ASSERT_EQ(2, stm.sampling_config().frequency());
  ASSERT_EQ(10240000u, stm.sampling_config().frequency_division());
  ASSERT_EQ(std::chrono::microseconds(500000), stm.sampling_config().period());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(10240000u, autd.link().stm_frequency_division(dev.idx()));
  }

  ASSERT_EQ(std::nullopt, stm.start_idx());
  ASSERT_EQ(std::nullopt, stm.finish_idx());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(-1, autd.link().stm_start_idx(dev.idx()));
    ASSERT_EQ(-1, autd.link().stm_finish_idx(dev.idx()));
  }

  stm.with_start_idx(0);
  ASSERT_TRUE(autd.send_async(stm).get());
  ASSERT_EQ(0, stm.start_idx());
  ASSERT_EQ(std::nullopt, stm.finish_idx());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(0, autd.link().stm_start_idx(dev.idx()));
    ASSERT_EQ(-1, autd.link().stm_finish_idx(dev.idx()));
  }

  stm.with_start_idx(std::nullopt);
  stm.with_finish_idx(0);
  ASSERT_TRUE(autd.send_async(stm).get());
  ASSERT_EQ(std::nullopt, stm.start_idx());
  ASSERT_EQ(0, stm.finish_idx());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(-1, autd.link().stm_start_idx(dev.idx()));
    ASSERT_EQ(0, autd.link().stm_finish_idx(dev.idx()));
  }

  stm = autd3::internal::GainSTM::from_sampling_config(autd3::internal::SamplingConfiguration::from_frequency_division(512))
            .add_gain(autd3::gain::Uniform(0x80))
            .add_gain(autd3::gain::Uniform(0x80));
  ASSERT_TRUE(autd.send_async(stm).get());
  ASSERT_EQ(20000.0, stm.frequency());
  ASSERT_EQ(2 * 20000.0, stm.sampling_config().frequency());
  ASSERT_EQ(512u, stm.sampling_config().frequency_division());
  ASSERT_EQ(std::chrono::microseconds(25), stm.sampling_config().period());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(512u, autd.link().stm_frequency_division(dev.idx()));
  }

  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(2u, autd.link().stm_cycle(dev.idx()));
    auto [intensities, phases] = autd.link().intensities_and_phases(dev.idx(), 0);
    ASSERT_TRUE(std::ranges::all_of(intensities, [](auto d) { return d == 0x80; }));
    ASSERT_TRUE(std::ranges::all_of(phases, [](auto p) { return p == 0; }));

    std::tie(intensities, phases) = autd.link().intensities_and_phases(dev.idx(), 1);
    ASSERT_TRUE(std::ranges::all_of(intensities, [](auto d) { return d == 0x80; }));
    ASSERT_TRUE(std::ranges::all_of(phases, [](auto p) { return p == 0; }));
  }

  stm.with_mode(autd3::internal::native_methods::GainSTMMode::PhaseFull);
  ASSERT_TRUE(autd.send_async(stm).get());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(2u, autd.link().stm_cycle(dev.idx()));
    auto [intensities, phases] = autd.link().intensities_and_phases(dev.idx(), 0);
    ASSERT_TRUE(std::ranges::all_of(intensities, [](auto d) { return d == 0xFF; }));
    ASSERT_TRUE(std::ranges::all_of(phases, [](auto p) { return p == 0; }));

    std::tie(intensities, phases) = autd.link().intensities_and_phases(dev.idx(), 1);
    ASSERT_TRUE(std::ranges::all_of(intensities, [](auto d) { return d == 0xFF; }));
    ASSERT_TRUE(std::ranges::all_of(phases, [](auto p) { return p == 0; }));
  }

  stm.with_mode(autd3::internal::native_methods::GainSTMMode::PhaseHalf);
  ASSERT_TRUE(autd.send_async(stm).get());
  for (const auto& dev : autd.geometry()) {
    ASSERT_EQ(2u, autd.link().stm_cycle(dev.idx()));
    auto [intensities, phases] = autd.link().intensities_and_phases(dev.idx(), 0);
    ASSERT_TRUE(std::ranges::all_of(intensities, [](auto d) { return d == 0xFF; }));
    ASSERT_TRUE(std::ranges::all_of(phases, [](auto p) { return p == 0; }));

    std::tie(intensities, phases) = autd.link().intensities_and_phases(dev.idx(), 1);
    ASSERT_TRUE(std::ranges::all_of(intensities, [](auto d) { return d == 0xFF; }));
    ASSERT_TRUE(std::ranges::all_of(phases, [](auto p) { return p == 0; }));
  }
}
